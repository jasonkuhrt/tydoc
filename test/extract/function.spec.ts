it('extracts docs', () => {
  expect(
    ctx.given(
      `
        export function a() {}
        export function b(a:boolean) {}
        export function c(a:string): number {}
      `
    )
  ).toMatchSnapshot()
})

it('treats a variable declaration initialized to a function as a function', () => {
  expect(
    ctx.given(`
      export const foo = function () {}
    `)
  ).toMatchSnapshot()
})

it('treats a variable declaration initialized to an arrow function as a function', () => {
  expect(
    ctx.given(`
      export const foo = () => {}
    `)
  ).toMatchSnapshot()
})

it('if function expression given name different than variable, is ignored', () => {
  expect(
    ctx.given(`
      export const foo = function foo2 () {}
    `).terms[0].name
  ).toEqual('foo')
})

describe('jsdoc', () => {
  it('is null when no jsDoc is present', () => {
    expect(
      ctx.given(
        `
          export function a() {}
        `
      ).terms[0].jsDoc
    ).toBeNull()
  })

  it('extracts doc from a variable statement', () => {
    expect(
      ctx.given(
        `
          /**
           * primary
           */
          export function a() {}
        `
      ).terms[0].jsDoc!.primary
    ).toMatchSnapshot()
  })

  it('splits multiple jsDoc blocks by primary and additional (closest to code is primary)', () => {
    expect(
      ctx.given(
        `
          /**
           * additional 2
           */
          /**
           * additional 1
           */
          /**
           * primary
           */
          export function a() {}
        `
      ).terms[0].jsDoc
    ).toMatchSnapshot()
  })

  it('whitespace and comments between multiple jsDoc blocks are ignored', () => {
    expect(
      ctx.given(
        `
          /**
           * additional 2
           */

           // boo 2

          /**
           * additional 1
           */

           // boo 1

          /**
           * primary
           */
          export function a() {}
        `
      ).terms[0].jsDoc
    ).toMatchSnapshot()
  })

  it.todo('keeps malformed jsdoc in raw form')
})
