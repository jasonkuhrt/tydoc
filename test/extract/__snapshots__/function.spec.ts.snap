// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extracts docs 1`] = `
Object {
  "hybrids": Array [],
  "length": 3,
  "terms": Array [
    Object {
      "jsDoc": null,
      "kind": "function",
      "languageLevel": "term",
      "name": "a",
      "signature": Object {
        "parameters": Array [],
        "return": Object {
          "name": "void",
        },
      },
      "sourceLocation": Object {
        "fileLine": 1,
        "filePath": "/test.ts",
      },
      "text": "export function a() {}",
    },
    Object {
      "jsDoc": null,
      "kind": "function",
      "languageLevel": "term",
      "name": "b",
      "signature": Object {
        "parameters": Array [
          Object {
            "name": "a",
            "type": Object {
              "name": "boolean",
            },
          },
        ],
        "return": Object {
          "name": "void",
        },
      },
      "sourceLocation": Object {
        "fileLine": 2,
        "filePath": "/test.ts",
      },
      "text": "export function b(a: boolean) {}",
    },
    Object {
      "jsDoc": null,
      "kind": "function",
      "languageLevel": "term",
      "name": "c",
      "signature": Object {
        "parameters": Array [
          Object {
            "name": "a",
            "type": Object {
              "name": "string",
            },
          },
        ],
        "return": Object {
          "name": "number",
        },
      },
      "sourceLocation": Object {
        "fileLine": 3,
        "filePath": "/test.ts",
      },
      "text": "export function c(a: string): number {}",
    },
  ],
  "types": Array [],
}
`;

exports[`jsdoc extracts doc from a variable statement 1`] = `
Object {
  "source": "/**
 * primary
 */",
}
`;

exports[`jsdoc splits multiple jsDoc blocks by primary and additional (closest to code is primary) 1`] = `
Object {
  "additional": Array [
    Object {
      "source": "/**
 * additional 2
 */",
    },
    Object {
      "source": "/**
 * additional 1
 */",
    },
  ],
  "primary": Object {
    "source": "/**
 * primary
 */",
  },
}
`;

exports[`jsdoc whitespace and comments between multiple jsDoc blocks are ignored 1`] = `
Object {
  "additional": Array [
    Object {
      "source": "/**
 * additional 2
 */",
    },
    Object {
      "source": "/**
 * additional 1
 */",
    },
  ],
  "primary": Object {
    "source": "/**
 * primary
 */",
  },
}
`;

exports[`treats a variable declaration initialized to a function as a function 1`] = `
Object {
  "hybrids": Array [],
  "length": 1,
  "terms": Array [
    Object {
      "jsDoc": null,
      "kind": "function",
      "languageLevel": "term",
      "name": "foo",
      "signature": Object {
        "parameters": Array [],
        "return": Object {
          "name": "void",
        },
      },
      "sourceLocation": Object {
        "fileLine": 1,
        "filePath": "/test.ts",
      },
      "text": "function() {}",
    },
  ],
  "types": Array [],
}
`;

exports[`treats a variable declaration initialized to an arrow function as a function 1`] = `
Object {
  "hybrids": Array [],
  "length": 1,
  "terms": Array [
    Object {
      "jsDoc": null,
      "kind": "function",
      "languageLevel": "term",
      "name": "foo",
      "signature": Object {
        "parameters": Array [],
        "return": Object {
          "name": "void",
        },
      },
      "sourceLocation": Object {
        "fileLine": 1,
        "filePath": "/test.ts",
      },
      "text": "() => {}",
    },
  ],
  "types": Array [],
}
`;
